README.md

```diff
@@ -28,86 +45,158 @@ These web servers include Rack handlers in their distributions:
+```ruby
+Grack::App.new(:adapter => Grack::RJGitAdapter)
+```
 
-Grack makes calls to the git binary through the GitAdapter abstraction class. Grack can be made to use a different backend by specifying a different Adapter class in Grack's configuration, for example:
+Alternative adapters available:
+* [rjgit_grack](http://github.com/dometto/rjgit_grack) lets Grack use the
+  [RJGit](http://github.com/repotag/rjgit) gem to implement Smart HTTP in pure
+  Jruby.
+
+### Developing Adapters
+
+Adapters are abstraction classes that handle the actual implementation of the
+Smart HTTP protocol (advertising refs, uploading and receiving packfiles). Such
+abstraction classes must have the following methods:
 
 ```ruby
-Grack::App.new({
-      :adapter => Grack::RJGitAdapter
-    })
+MyAdapter.repository_path=(repository_path)
+MyAdapter.exist?
+MyAdapter.handle_pack(kind, io_in, io_out, opts = {})
+MyAdapter.file(path)
+MyAdapter.update_server_info
+MyAdapter.allow_push?
+MyAdapter.allow_pull?
 ```
 
-Alternative adapters available:
-- [rjgit_grack](http://github.com/dometto/rjgit_grack) lets Grack use the [RJGit](http://github.com/repotag/rjgit) gem to implement smart-http in pure jruby.
+See `Grack::GitAdapter` for more detailed documentation and an example
+implementation.
 
-See below if you are looking to create a custom Adapter.
+## Features
 
-Contributing
-========================
-If you would like to contribute to the Grack project, I prefer to get
-pull-requests via GitHub.  You should include tests for whatever functionality
-you add.  Just fork this project, push your changes to your fork and click
-the 'pull request' button.
+* Supports Git Smart HTTP protocol.
+* Supports Git Basic HTTP protocol.
+* Limits push/pull access globally and per-repository.
+* Thread safe operation.
 
-Run 'bundle install' to install development dependencies. Then you should be able to run the tests with a 'rake' command. On ruby >= 1.9, a coverage report will be generated using simplecov. On ruby 1.8, use rcov instead: uncomment the relevant line in the Gemfile and use 'rake rcov'. 
+## Known Bugs/Limitations
 
-### Developing Adapters
+* Will likely block fully evented web servers when using the stock Git adapter.
+
+## Synopsis
 
-Adapters are abstraction classes that handle the actual implementation of the smart-http protocol (advertising refs, uploading and receiving packfiles). Such abstraction classes must have the following methods:
+In `config.ru`:
 
 ```ruby
-MyAdapter.receive_pack(repository_path, opts = {}, &block)
-MyAdapter.upload_pack(repository_path, opts = {}, &block)
-MyAdapter.update_server_info(repository_path, opts = {}, &block) # The equivalent of 'git update-server-info'. Optional, for falling back to dumb-http mode.
-MyAdapter.get_config_setting(repository_path, key) # Always returns a string, e.g. "false" for key "core.bare".
+require 'grack/app'
+require 'grack/git_adapter_factory'
+
+config = {
+  :root => '/path/to/bare/repositories',
```

Comment:
Maybe it's enough to have a note in the readme or in the wiki about non-bare repos. (For most use cases, bare should be encouraged anyway, because recent versions of git will refuse to push to non-bare repositories by default)